tezos/api/src/ocaml_conv/to_ocaml.rs

```diff
@@ -0,0 +1,252 @@
+    ApplyBlockRequest, BeginConstructionRequest, FfiRpcService, JsonRpcRequest,
+    PrevalidatorWrapper, ProtocolJsonRpcRequest, ValidateOperationRequest,
+};
+use crypto::hash::{BlockHash, ContextHash, Hash, OperationListListHash, ProtocolHash};
+use tezos_messages::p2p::encoding::prelude::{BlockHeader, Operation};
+use znfe::{
+    ocaml, OCamlInt, OCaml, OCamlAllocResult, OCamlAllocToken, OCamlBytes, OCamlInt32, OCamlInt64,
+    OCamlList, ToOCaml,
+};
+
+// Headers
+struct BlockHeaderShellHeader {}
+struct OperationShellHeader {}
+
+ocaml! {
+    // Requests
+
+    alloc fn alloc_apply_block_request(
+        chain_id: OCamlBytes,
+        block_header: BlockHeader,
+        pred_header: BlockHeader,
+        max_operations_ttl: OCamlInt,
+        operations: OCamlList<OCamlList<Operation>>,
+    ) -> ApplyBlockRequest;
+
+    alloc fn alloc_begin_construction_request(
+        chain_id: OCamlBytes,
+        predecessor: BlockHeader,
+        protocol_data: Option<OCamlBytes>,
+    ) -> BeginConstructionRequest;
+
+    alloc fn alloc_validate_operation_request(
+        prevalidator: PrevalidatorWrapper,
+        operation: Operation,
+    ) -> ValidateOperationRequest;
+
+    alloc fn alloc_json_rpc_request(
+        body: OCamlBytes,
+        context_path: OCamlBytes,
+    ) -> JsonRpcRequest;
+
+    alloc fn alloc_protocol_json_rpc_request(
+        block_header: BlockHeader,
+        chain_id: OCamlBytes,
+        chain_arg: OCamlBytes,
+        request: JsonRpcRequest,
+        ffi_service: FfiRpcService,
+    ) -> ProtocolJsonRpcRequest;
+
+    // Headers
+
+    alloc fn alloc_block_header_shell_header(
+        level: OCamlInt32,
+        proto_level: OCamlInt,
+        predecessor: OCamlBlockHash,
+        timestamp: OCamlInt64,
+        validation_passes: OCamlInt,
+        operations_hash: OCamlOperationListListHash,
+        fitness: OCamlList<OCamlBytes>,
+        context: OCamlContextHash,
+    ) -> BlockHeaderShellHeader;
+
+    alloc fn alloc_block_header(
+        shell: BlockHeaderShellHeader,
+        protocol_data: OCamlBytes,
+    ) -> BlockHeader;
+
+    alloc fn alloc_operation_shell_header(
+        branch: OCamlBlockHash,
+    ) -> OperationShellHeader;
+
+    // Operation
+
+    alloc fn alloc_operation(
+        shell: OperationShellHeader,
+        proto: OCamlBytes,
+    ) -> Operation;
+
+    // Wrappers
+
+    alloc fn alloc_prevalidator_wrapper(
+        chain_id: OCamlBytes,
+        protocol: OCamlProtocolHash,
+    ) -> PrevalidatorWrapper;
+
+    // Hashes
+
+    alloc fn alloc_operation_list_list_hash(hash: OCamlBytes) -> OCamlOperationListListHash;
+    alloc fn alloc_operation_hash(hash: OCamlBytes) -> OCamlOperationHash;
+    alloc fn alloc_block_hash(hash: OCamlBytes) -> OCamlBlockHash;
+    alloc fn alloc_context_hash(hash: OCamlBytes) -> OCamlContextHash;
+    alloc fn alloc_protocol_hash(hash: OCamlBytes) -> OCamlProtocolHash;
+}
+
+unsafe impl ToOCaml<ApplyBlockRequest> for ApplyBlockRequest {
```

Comment:
@tizoc 
Do we have all this converions/roundtrips catched in test_bytes_roundtrips.rs? 
It would be very usefull to have all this conversions catched in some test to be sure it works.
We would like to catch any incompatibility between structs (rust vs ocaml)  at least at tests.
What would you recommend?

Reply:
I think most are either directly or indirectly being converted during the tests (same as when the values where being encoded/decoded), but let me doublecheck.

When I asked some time ago how often these structs changed it was because I was worrying about this. I agree that it should be validated somehow in tests, but I haven't figured out the best way to do it yet. I will think about it more.

Reply:
I agree, they are tested indirectly in other tests.
This is important part "border between rust/ocaml", so I think it deserves direct dedicated tests :)
I am ok, if just test_bytes_roundtrips.rs covers that